sched/wqueue/kwork_thread.c

```diff
@@ -169,19 +168,20 @@ static int work_thread(int argc, FAR char *argv[])
           /* Mark the work as no longer being queued */
 
           work->worker = NULL;
+          leave_critical_section(flags);
 
           /* Do the work.  Re-enable interrupts while the work is being
            * performed... we don't have any idea how long this will take!
            */
 
-          leave_critical_section(flags);
           CALL_WORKER(worker, arg);
-          flags = enter_critical_section();
+        }
+      else
+        {
+          leave_critical_section(flags);
```

Comment:
why do we need changes in this file? I see only function reordering, but no functional changes

Reply:
Now the order:
while (1)
{
  enter_critical_section();
  nxsem_wait_uninterruptible();
  ...
  // access the work_s with critical_section
  ...
  leave_critical_section();
}

The restore will make sure we are back to critical_section state, but it is different to understand for a new reader.

After modify:

while (1)
{
  nxsem_wait_uninterruptible();
  enter_critical_section();
  ...
  // access the work_s with critical_section
  ...
  leave_critical_section();
}

This is easy to understand.

Reply:
Again, I do not see any issue with the current approach. The existing code ensures that all code except `CALL_WORKER(worker, arg);` is executed inside the critical section, It is straight forward and easy to understand at least to me.
I will let other reviewers to write their opinion on this.

Reply:
Again and again, I haven't say that there is an issue.

The current code:
The restore will make sure we are back to critical_section state, but it is different to understand for a new reader.

The new code:
This is easy to understand.